// node-opcua-pki/dist/test/test_concurrency.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = __importDefault(require("path"));
require("should");
const lib_1 = require("../lib");
const helpers_1 = require("./helpers");
describe("Concurrency", function () {
    const testData = (0, helpers_1.beforeTest)(this);
    it("should not crash if multiple CertificateManager instances share the same folder", () => __awaiter(this, void 0, void 0, function* () {
        const location = path_1.default.join(testData.tmpFolder, "concurrency");
        const workWithCertificateManager = (n) => __awaiter(this, void 0, void 0, function* () {
            const cm1 = new lib_1.CertificateManager({
                location,
            });
            console.log("starting ", n);
            yield cm1.initialize();
            const now = new Date();
            const endDate = new Date(now.getFullYear() + 7, 10, 10);
            const duration = 10000;
            const params = {
                applicationUri: "MY:APPLICATION:URI",
                dns: ["some.other.domain.com", "my.domain.com"],
                ip: ["192.123.145.121"],
                subject: "/CN=MyCommonName",
                // can only be TODAY due to openssl limitation : startDate: new Date(2010,2,2),
                validity: duration,
                startDate: now,
            };
            yield cm1.createSelfSignedCertificate(params);
            console.log("ah!", n);
        });
        const promises = [
            workWithCertificateManager(1),
            workWithCertificateManager(2),
            workWithCertificateManager(3),
            workWithCertificateManager(4),
        ];
        yield Promise.all(promises);
    }));
});
