// node-opcua-pki/dist/test/test_create_signing_request.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = __importDefault(require("path"));
const fs_1 = __importDefault(require("fs"));
const without_openssl_1 = require("../lib/toolbox/without_openssl");
const create_certificate_signing_request_1 = require("../lib/toolbox/with_openssl/create_certificate_signing_request");
const certificate_manager_1 = require("../lib/pki/certificate_manager");
const fs_2 = require("fs");
const node_opcua_crypto_1 = require("node-opcua-crypto");
const helpers_1 = require("./helpers");
const ca_1 = require("../lib/ca");
describe("comparing two implementations of createCertificateSigningRequestAsync", function () {
    const rootDir = path_1.default.join(__dirname, "../tmp/certificates/PKI-2");
    const configFile = path_1.default.join(rootDir, "own/openssl.cnf");
    const privateKey = path_1.default.join(rootDir, "own/private/private_key.pem");
    before(() => __awaiter(this, void 0, void 0, function* () {
        const options = {
            location: rootDir,
        };
        (0, fs_2.mkdirSync)(rootDir, { recursive: true });
        const cm = new certificate_manager_1.CertificateManager(options);
        yield cm.initialize();
    }));
    it("should product identical results (or sufficiently identical)", () => __awaiter(this, void 0, void 0, function* () {
        const applicationUri = "urn:localhost:MyProduct";
        const dns = ["localhost", "my.domain.com"];
        const ip = ["192.168.1.1"];
        const subject = "/C=FR/ST=IDF/L=Paris/O=MyOrganization/OU=MyDepartment/CN=MyCommonName";
        const csr1 = path_1.default.join(__dirname, "../tmp/without_openssl.csr");
        const csr2 = path_1.default.join(__dirname, "../tmp/with_openssl.csr");
        yield (0, without_openssl_1.createCertificateSigningRequestAsync)(csr1, {
            rootDir,
            configFile,
            privateKey,
            applicationUri,
            dns,
            ip,
            subject,
            purpose: node_opcua_crypto_1.CertificatePurpose.ForApplication,
        });
        yield (0, create_certificate_signing_request_1.createCertificateSigningRequestWithOpenSSL)(csr2, {
            rootDir,
            configFile,
            privateKey,
            applicationUri,
            dns,
            ip,
            subject,
            purpose: node_opcua_crypto_1.CertificatePurpose.ForApplication,
        });
        //  openssl req -text -noout -verify -in server.csr
    }));
    describe("createCertificateSigningRequestAsync", () => {
        let theCertificateAuthority;
        before(() => __awaiter(this, void 0, void 0, function* () {
            const testData = (0, helpers_1.beforeTest)(this);
            theCertificateAuthority = new ca_1.CertificateAuthority({
                keySize: 2048,
                location: path_1.default.join(testData.tmpFolder, "CA"),
            });
            yield theCertificateAuthority.initialize();
        }));
        it("should be possible to create a certificate from a createCertificateSigningRequestAsync  created without openssl )", () => __awaiter(this, void 0, void 0, function* () {
            const applicationUri = "urn:localhost:MyProduct";
            const dns = ["localhost", "my.domain.com"];
            const ip = ["192.168.1.1"];
            const subject = "/C=FR/ST=IDF/L=Paris/O=MyOrganization/OU=MyDepartment/CN=MyCommonName";
            const certificateSigningRequestFilename = path_1.default.join(__dirname, "../tmp/without_openssl.csr");
            yield (0, without_openssl_1.createCertificateSigningRequestAsync)(certificateSigningRequestFilename, {
                rootDir,
                configFile,
                privateKey,
                applicationUri,
                dns,
                ip,
                subject,
                purpose: node_opcua_crypto_1.CertificatePurpose.ForApplication,
            });
            const params = {
                applicationUri: "BAD SHOULD BE IN REQUEST",
                startDate: new Date(2011, 25, 12),
                validity: 10 * 365,
            };
            const certificateFilename = path_1.default.join(rootDir, "cert1.pem");
            if (fs_1.default.existsSync(certificateFilename)) {
                fs_1.default.unlinkSync(certificateFilename);
            }
            yield theCertificateAuthority.signCertificateRequest(certificateFilename, certificateSigningRequestFilename, params);
        }));
    });
});
