// node-opcua-nodeset-ua/dist/enum_port_id_subtype.js
"use strict";
// ----- this file has been automatically generated - do not edit
Object.defineProperty(exports, "__esModule", { value: true });
exports.EnumPortIdSubtype = void 0;
/**
 * |           |                                                            |
 * |-----------|------------------------------------------------------------|
 * | namespace |http://opcfoundation.org/UA/                                |
 * | nodeClass |DataType                                                    |
 * | name      |PortIdSubtype                                               |
 * | isAbstract|false                                                       |
 */
var EnumPortIdSubtype;
(function (EnumPortIdSubtype) {
    /**
     * Represents a port identifier based on the ifAlias
     * MIB object defined in IETF RFC 2863.
     */
    EnumPortIdSubtype[EnumPortIdSubtype["InterfaceAlias"] = 1] = "InterfaceAlias";
    /**
     * Represents a port identifier based on the value
     * of entPhysicalAlias (defined in IETF RFC 2737)
     * for a port component (i.e., entPhysicalClass
     * value of port(10) or backplane(4)), within the
     * containing chassis.
     */
    EnumPortIdSubtype[EnumPortIdSubtype["PortComponent"] = 2] = "PortComponent";
    /**
     * Represents a port identifier based on a unicast
     * source address (encoded in network byte order and
     * IEEE 802.3 canonical bit order) which has been
     * detected by the agent and associated with a
     * particular port (IEEE Std 802-2014).
     */
    EnumPortIdSubtype[EnumPortIdSubtype["MacAddress"] = 3] = "MacAddress";
    /**
     * Represents a port identifier based on a network
     * address, detected by the agent and associated
     * with a particular port.
     */
    EnumPortIdSubtype[EnumPortIdSubtype["NetworkAddress"] = 4] = "NetworkAddress";
    /**
     * Represents a port identifier based on the ifName
     * MIB object, defined in IETF RFC 2863.
     */
    EnumPortIdSubtype[EnumPortIdSubtype["InterfaceName"] = 5] = "InterfaceName";
    /**
     * Represents a port identifier based on the
     * agent-local identifier of the circuit (defined in
     * IETF RFC 3046), detected by the agent and
     * associated with a particular port.
     */
    EnumPortIdSubtype[EnumPortIdSubtype["AgentCircuitId"] = 6] = "AgentCircuitId";
    /**
     * Represents a port identifier based on a value
     * locally assigned.
     */
    EnumPortIdSubtype[EnumPortIdSubtype["Local"] = 7] = "Local";
})(EnumPortIdSubtype || (exports.EnumPortIdSubtype = EnumPortIdSubtype = {}));
